### PR TITLE
AEJ-1620: Follow best practices when using primitive data types

### PR DESCRIPTION
**Pull request by LiquidLemon**
I added comments to all instances of typecasting I could find. I don't feel like this is a valuable change but it is explicitly requested in the ticket.

All of the other typecasts I could find had to do with casting pointers. I decided to leave those out because they're more straightforward and don't do anything at runtime.

### CODE REVIEW

File: src/main/native/compression/IntelDeflater.cc [hunk 1]
```diff
@@ -323,7 +323,8 @@ JNIEXPORT jint JNICALL Java_com_intel_gkl_compression_IntelDeflater_deflateNativ
     }
 
     // return number of bytes written to output buffer
-    long bytes_out = outputBufferLength - lz_stream->avail_out;
+    // ISAL is capable of handling 64 bit buffer sizes but we're truncating to stay consistent with the Java interface.
```
Comment (Kmannth) [inline src/main/native/compression/IntelDeflater.cc:326]: Please check 64 is the right value in the comment.

File: src/main/native/smithwaterman/PairWiseSW.h [hunk 2]
```diff
@@ -417,7 +423,8 @@ void inline getCIGAR(SeqPair *p, int16_t *cigarBuf_, int32_t tid)
                    curSize += snprintf(p->cigar + curSize, expectedLength + 1, "%d%c", cigarArray[2 * i + 1], state);
            }
     }
-    p->cigarCount = strnlen(p->cigar, curSize);
+    // The length of the string can be truncated to int16_t because it will never be larger than the at most 16 bit long input.
+    p->cigarCount = (int16_t)strnlen(p->cigar, curSize);
```
Comment (Kmannth) [inline src/main/native/smithwaterman/PairWiseSW.h:427]: Note to me.  Review.

File: src/main/native/smithwaterman/PairWiseSW.h [hunk 3]
```diff
@@ -452,8 +459,9 @@ int32_t CONCAT(runSWOnePairBT_,SIMD_ENGINE)(int32_t match, int32_t mismatch, int
     SeqPair p;
     p.seq1 = seq1;
     p.seq2 = seq2;
-    p.len1 = len1;
-    p.len2 = len2;
+    // len1 and len2 are bounded by the length of the input with is at most 16 bit.
+    p.len1 = (int16_t)len1;
```
Comment (Kmannth) [inline src/main/native/smithwaterman/PairWiseSW.h:463]: Can we look to change " int32_t len1, int32_t len2" to int16 types to align with the 16 bit Max values reported.

File: src/main/native/compression/IntelDeflater.cc (hunk 1 above)
Comment (LiquidLemon) [reply src/main/native/compression/IntelDeflater.cc]: I removed this comment completely as no type conversions are actually being performed here.

File: src/main/native/smithwaterman/PairWiseSW.h (hunk 3 above)
Comment (LiquidLemon) [reply src/main/native/smithwaterman/PairWiseSW.h]: Done. I also did a similar thing with the strategy value coming from Java.

File: src/main/native/smithwaterman/PairWiseSW.h (hunk 2 above)
Comment (Kmannth) [reply src/main/native/smithwaterman/PairWiseSW.h:427]: Last issue to review/resolve.

Comment (Kmannth) [reply src/main/native/smithwaterman/PairWiseSW.h:427]: According to SAM file spect this should be uint32. https://samtools.github.io/hts-specs/SAMv1.pdf  section 4.2 pg 15. uint32 t[n cigar op]

**Comment by LiquidLemon** [comment]
I seem to have fixed all the issues that came up when rebasing. Compilation and tests work now and I have manually reviewed the list of changes to make sure that I didn't leave anything out.

**Review by Kmannth — APPROVED** [review]

**Pull request merged**